### PR TITLE
[ALLUXIO-1882] Enable using EC2 Instance profile for S3 Access

### PR DESCRIPTION
**Pull request by csenol**
[ALLUXIO-1882](https://alluxio.atlassian.net/browse/ALLUXIO-1882)

### CODE REVIEW

**Comment by calvinjia** [comment]
Jenkins, ok to test.

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java [hunk 1]
```diff
@@ -79,17 +78,13 @@
    *
    * @param bucketName the name of the bucket
    * @param conf the configuration for Alluxio
+   * @param awsCredentials AWS Credentials configuration for S3 Access
    * @throws ServiceException when a connection to S3 could not be created
    */
-  public S3UnderFileSystem(String bucketName, Configuration conf) throws ServiceException {
+  public S3UnderFileSystem(String bucketName, Configuration conf,
+                           AWSCredentials awsCredentials) throws ServiceException {
```
Comment (calvinjia) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: Could you make the formatting consistent with the codebase here? (args one line, new line for throws)

**Comment by calvinjia** [comment]
@csenol Thanks for implementing this feature! Could you add the appropriate documentation on how to use it (`docs/en/Configuring-Alluxio-with-S3.md`) as well as a unit test if possible?

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystemFactory.java [hunk 2]
```diff
@@ -39,20 +39,23 @@
...
-      try {
-        return new S3UnderFileSystem(uri.getHost(), configuration);
-      } catch (ServiceException e) {
-        LOG.error("Failed to create S3UnderFileSystem.", e);
-        throw Throwables.propagate(e);
-      }
+      creds = new AWSCredentials(configuration.get(Constants.S3_ACCESS_KEY),
+              configuration.get(Constants.S3_SECRET_KEY));
+    } else {
+      creds = AWSEC2IAMSessionCredentials.loadFromEC2InstanceData(true);
```
Comment (gpang) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystemFactory.java]: What would happen if the AWS credentials are not specified, and the application is NOT running on EC2? It seems like there is one (error) case not handled here?

**Comment by gpang** [comment]
@csenol thanks for the improvement! I had a minor question.

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystemFactory.java (hunk 2 above)
Comment (csenol) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystemFactory.java]: How should we handle this case ? Do you think if it makes sense to introduce a new config/flag which is something like `fs.s3n.useInstanceRole=true` ? If this is set to false(which is by default) and instance is not running on AWS or running without a role, we can throw an exception saying `neither creds nor role provided`. Does this approach makes sense ?

Comment (gpang) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystemFactory.java]: What happens when `AWSEC2IAMSessionCredentials.loadFromEC2InstanceData()` is called when NOT on ec2? Does it return null? Does it throw and exception?

Comment (csenol) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystemFactory.java]: It throws an exception and I propagated it

**Comment by csenol** [comment]
@calvinjia I added a minimal set of documentation.

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java [hunk 3]
```diff
@@ -80,18 +79,13 @@
    *
    * @param uri the {@link AlluxioURI} for this UFS
    * @param conf the configuration for Alluxio
+   * @param awsCredentials AWS Credentials configuration for S3 Access
    * @throws ServiceException when a connection to S3 could not be created
    */
-  public S3UnderFileSystem(AlluxioURI uri, Configuration conf) throws ServiceException {
+  public S3UnderFileSystem(AlluxioURI uri, Configuration conf, AWSCredentials awsCredentials)
+          throws ServiceException {
```
Comment (gpang) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: I think this is indented too much?

**Comment by gpang** [comment]
I left a minor comment, but this looks good. Thanks for the improvement!

**Comment by gpang** [comment]
LGTM

**Comment by apc999** [comment]
LGTM

**Comment by calvinjia** [comment]
LGTM

**Comment by oreh** [comment]
When would this be included in the formal release? This is an important update to help us deploy Alluxio in EC2.

**Pull request merged**